src/main/java/picard/sam/SamErrorMetric/CollectSamErrorMetrics.java

```diff
@@ -328,30 +416,205 @@ private void writeMetricsFileForAggregator(final BaseErrorAggregation locusAggre
      */
     private void addLocusBases(final Collection<BaseErrorAggregation> aggregatorList, final SAMLocusAndReference info) {
-        info.getRecordAndOffsets()
-                .forEach(rao -> aggregatorList
-                        .forEach(l -> l.addBase(rao, info)));
+        // Matching bases
+        for(SamLocusIterator.RecordAndOffset rao : info.getRecordAndOffsets()) {
+            addAndFilterRecordAndOffset(aggregatorList, rao, info, BaseOperation.Match);
+        }
+
+        // Deleted bases
+        for (SamLocusIterator.RecordAndOffset deletionRao : info.getLocus().getDeletedInRecord()) {
+            addAndFilterRecordAndOffset(aggregatorList, deletionRao, info, BaseOperation.Deletion);
+        }
+
+        // Inserted bases
+        for (SamLocusIterator.RecordAndOffset insertionRao : info.getLocus().getInsertedInRecord()) {
+            addAndFilterRecordAndOffset(aggregatorList, insertionRao, info, BaseOperation.Insertion);
+        }
+    }
+
+    /**
+     * Gets a unique identifier for a read. In BAM files, this identifier is guaranteed to be unique, however, in SAM
+     * files, this identifier is not available and the read name is used instead. This is likely to cause collisions
+     * and as a consequence not all matching reads might be written to the output. To avoid this, consider using a BAM file.
+     * @param read The read to calculate the unique Id for
+     * @return A String representing the position of the read in the BAM file, thereby acting as a unique identifier.
+     *         For SAM files, the read name is used instead.
+     */
+    private String getUniqueReadId(final SAMRecord read) {
+        if (read.getFileSource() == null) {
+            log.warn("There is no supported way to generate a unique read ID for SAM files. Instead, the " +
+                    "read name is used, which is likely to cause collisions. As a consequence, not all reads matching " +
+                    "the filter criteria might be written to the output. To avoid this, consider using a BAM file.");
+            return read.getReadName();
+        }
+        return String.valueOf(((BAMFileSpan)read.getFileSource().getFilePointer()).getFirstOffset());
+    }
+
+    /**
+     * Adds a read to a list of reads to be written to output at the end of the error collection. For multiple matches,
+     * a counter is increased to avoid multiple output of the same read
+     * @param read Read to be added to the output list
+     */
+    private void addOutputRead(final SAMRecord read) {
+        String readId = getUniqueReadId(read);
+        if (outputReadOccurrence.containsKey(readId)) {
+            outputReadOccurrence.put(readId, outputReadOccurrence.get(readId) + 1);
+        }
+        else {
+            outputReadOccurrence.put(readId, 1);
+        }
+
+        if (!outputReadsById.containsKey(readId)) {
+            outputReadsById.put(readId, read);
```

Comment:
you're going to store all the reads in memory? this seems a bit much!

Reply:
Quite embarrassing mistake really. Fixed it so that reads are written to file immediately.